File: src/main/java/macro/library/open_library/OpenLibrary.java
package macro.library.open_library;

import macro.library.Isbn;
import macro.library.Util;
import macro.library.book.Book;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONObject;

/**
 * Created by Macro303 on 2019-Oct-22
 */
public abstract class OpenLibrary {
	@NotNull
	private static final String URL = "http://openlibrary.org/api/books";

	@Nullable
	public static Book searchBook(@NotNull Isbn isbn) {
		var url = URL + "?bibkeys=ISBN:" + isbn + "&format=json&jscmd=data";
		var request = Util.httpRequest(url);
		if (request == null)
			return null;
		JSONObject response = request.getObject();
		var bookObj = response.optJSONObject("ISBN:" + isbn);
		if (bookObj == null)
			return null;
		var title = bookObj.getString("title");
		var subtitle = bookObj.optString("subtitle");
		if (subtitle.isBlank())
			subtitle = null;
		var authors = bookObj.optJSONArray("authors");
		StringBuilder authorStr = new StringBuilder();
		for (var author : authors) {
			if (authorStr.length() != 0)
				authorStr.append(";");
			authorStr.append(((JSONObject) author).getString("name"));
		}
		var publishers = bookObj.optJSONArray("publishers");
		StringBuilder publisherStr = new StringBuilder();
		for (var publisher : publishers) {
			if (publisherStr.length() != 0)
				publisherStr.append(";");
			publisherStr.append(((JSONObject) publisher).getString("name"));
		}
		return new Book(isbn, title, subtitle, authorStr.toString(), publisherStr.toString());
	}
}

File: src/main/kotlin/github/macro/external/OpenLibrary.kt
package github.macro.external

import github.macro.Utils
import github.macro.book.Book
import kong.unirest.json.JSONArray
import kong.unirest.json.JSONObject
import github.macro.book.Isbn
import org.apache.logging.log4j.LogManager

/**
 * Created by Macro303 on 2019-Nov-12
 */
object OpenLibrary {
	private val LOGGER = LogManager.getLogger(OpenLibrary::class.java)
	private val URL = "http://openlibrary.org/api/books"

	fun searchBook(isbn: Isbn): Book? {
		val url = "$URL?bibkeys=ISBN:$isbn&format=json&jscmd=data"
		val request = Utils.httpRequest(url) ?: return null
		val response = request.getObject()
		val bookObj = response.optJSONObject("ISBN:$isbn") ?: return null
		val title = bookObj.getString("title")
		var subtitle: String? = bookObj.optString("subtitle")
		if (subtitle.isNullOrBlank())
			subtitle = null
		var publishers: JSONArray? = bookObj.optJSONArray("publishers")
		if (publishers == null)
			publishers = JSONArray()
		val publisherStr = StringBuilder()
		for (publisher in publishers) {
			if (publisherStr.isNotEmpty())
				publisherStr.append(";")
			publisherStr.append((publisher as JSONObject).getString("name"))
		}
		val idObj = bookObj.getJSONObject("identifiers")
		val googleBooksId = idObj.optJSONArray("google")?.map { it?.toString() }?.firstOrNull()
		val goodreadsId = idObj.optJSONArray("goodreads")?.map { it?.toString() }?.firstOrNull()
		val openLibraryId = idObj.optJSONArray("openlibrary")?.map { it?.toString() }?.firstOrNull()
		return Book(
			isbn = isbn,
			title = title,
			subtitle = subtitle,
			publisher = publisherStr.toString(),
			googleBooksId = googleBooksId,
			goodreadsId = goodreadsId,
			openLibraryId = openLibraryId
		).add()
	}

	fun selectBook(book: Book): Book {
		return book
	}
}